### PR TITLE
Generalize daemon configuration provisioning 

### PR DESCRIPTION
**Pull request by changweige**
Refactor how daemon configuration file or string is provided to nydusd using a generic method.
Reduce times of loading configurations from disk and clean code.

In addition, move blob manager preheating logics to new packages from filesystem package and compose a e2e test case for it.

### CODE REVIEW

File: snapshot/snapshot.go [hunk 1]
```diff
@@ -119,20 +126,34 @@ func NewSnapshotter(ctx context.Context, cfg *config.Config) (snapshots.Snapshot
...
+	hasDaemon := cfg.DaemonMode != config.DaemonModeNone
 
 	nydusFs, err := fspkg.NewFileSystem(ctx, opts...)
 	if err != nil {
 		return nil, errors.Wrap(err, "failed to initialize nydus filesystem")
 	}
 
+	var blobMgr *blob.Manager
+	if fuseConfig, ok := daemonConfig.(*daemonconfig.FuseDaemonConfig); ok {
+		if fuseConfig.Device.Backend.BackendType == "localfs" &&
```
Comment (imeoer) [inline snapshot/snapshot.go:145]: Maybe give a comment to describe what's doing here.

File: pkg/daemon/daemon.go [hunk 2]
```diff
@@ -200,12 +195,13 @@ func (d *Daemon) SharedMount() error {
 		return err
 	}
 
+	cfg, _ := d.Config.DumpString()
```
[the remark this answers is not in the document]
Comment (changweige) [reply pkg/daemon/daemon.go]: fixed

File: pkg/filesystem/fs/fs.go [hunk 3]
```diff
@@ -544,23 +459,49 @@ func (fs *Filesystem) Mount(snapshotID string, labels map[string]string) (err er
...
+	// manager knows where to find those blobs.
+	cacheDir := fs.cacheMgr.CacheDir()
+
+	params := map[string]string{
+		daemonconfig.Bootstrap: bootstrap,
+		// FIXME: Does nydusd really stores cache files here?
+		daemonconfig.WorkDir:  workDir,
+		daemonconfig.CacheDir: cacheDir}
+
+	daemonconfig.SupplementDaemonConfig(cfg, imageID, d.SnapshotID, false, labels, params)
```
[the remark this answers is not in the document]
Comment (changweige) [reply pkg/filesystem/fs/fs.go]: fixed

File: pkg/filesystem/fs/fs.go [hunk 4]
```diff
@@ -734,14 +654,19 @@ func (fs *Filesystem) initSharedDaemon() (err error) {
...
 			}
 		}
 	}()
 
-	if err := fs.manager.StartDaemon(d); err != nil {
-		return errors.Wrap(err, "failed to start shared daemon")
+	// Shared nydusd daemon does not need configuration to start process but
+	// it is loaded when requesting mount api
+	d.Config = fs.Manager.DaemonConfig
+	d.Config.DumpFile(d.ConfigFile())
```
[the remark this answers is not in the document]
Comment (changweige) [reply pkg/filesystem/fs/fs.go]: fixed

File: pkg/filesystem/fs/fs.go [hunk 5]
```diff
@@ -673,37 +613,17 @@ func (fs *Filesystem) BootstrapFile(id string) (string, error) {
...
-
-	if fs.cacheMgr != nil {
-		// Overriding work_dir option of nydusd config as we want to set it
-		// via snapshotter config option to let snapshotter handle blob cache GC.
-		cfg.Device.Cache.Config.WorkDir = fs.cacheMgr.CacheDir()
-	}
-	return cfg, nil
-}
+func (fs *Filesystem) mount(d *daemon.Daemon) error {
+	d.Config.DumpFile(d.ConfigDir)
```
[the remark this answers is not in the document]
Comment (changweige) [reply pkg/filesystem/fs/fs.go]: fixed

File: snapshot/snapshot.go (hunk 1 above)
Comment (changweige) [reply snapshot/snapshot.go:147]: Comments added

File: config/daemonconfig/fscache.go [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+func LoadFscacheConfig(p string) (*FscacheDaemonConfig, error) {
+	var cfg FscacheDaemonConfig
+	b, err := os.ReadFile(p)
+	if err != nil {
+		return nil, errors.Wrapf(err, "read fscache configuration file %s", p)
+	}
+	if err = json.Unmarshal(b, &cfg); err != nil {
+		return nil, errors.Wrapf(err, "unmarshal")
+	}
+	return &cfg, nil
```
[the remark this answers is not in the document]
Comment (changweige) [reply config/daemonconfig/fscache.go:61]: Added the validation

**Review by imeoer — APPROVED** [review]
LGTM!

**Pull request merged**